tempodb/blocklist/poller.go

```diff
@@ -151,44 +154,90 @@ func (p *Poller) Do(previous *List) (PerTenant, PerTenantCompacted, error) {
 
-	consecutiveErrors := 0
+		finalErr       atomic.Error
+		tenantFailures atomic.Int32
+
+		tenantFailuresRemaining = p.cfg.TolerateTenantFailures
+	)
 
 	for _, tenantID := range tenants {
-		newBlockList, newCompactedBlockList, err := p.pollTenantAndCreateIndex(ctx, tenantID, previous)
-		if err != nil {
-			level.Error(p.logger).Log("msg", "failed to poll or create index for tenant", "tenant", tenantID, "err", err)
-			consecutiveErrors++
-			if consecutiveErrors > p.cfg.TolerateConsecutiveErrors {
-				level.Error(p.logger).Log("msg", "exiting polling loop early because too many errors", "errCount", consecutiveErrors)
-				return nil, nil, err
+		// Exit early if we have exceeded our tolerance for number of failing tenants.
+		if tenantFailures.Load() >= int32(tenantFailuresRemaining) {
+			if lastErr := finalErr.Load(); lastErr != nil {
+				level.Error(p.logger).Log("msg", "exiting polling loop early because too many errors")
+				return nil, nil, lastErr
 			}
-
-			blocklist[tenantID] = previous.Metas(tenantID)
-			compactedBlocklist[tenantID] = previous.CompactedMetas(tenantID)
-			continue
 		}
 
-		consecutiveErrors = 0
-		if len(newBlockList) > 0 || len(newCompactedBlockList) > 0 {
-			blocklist[tenantID] = newBlockList
-			compactedBlocklist[tenantID] = newCompactedBlockList
+		wg.Add(1)
+		go func(tenantID string) {
+			defer wg.Done()
 
-			metricBlocklistLength.WithLabelValues(tenantID).Set(float64(len(newBlockList)))
+			var (
+				consecutiveErrorsRemaining = p.cfg.TolerateConsecutiveErrors
+				newBlockList               = make([]*backend.BlockMeta, 0)
+				newCompactedBlockList      = make([]*backend.CompactedBlockMeta, 0)
+				err                        error
+			)
 
-			backendMetaMetrics := sumTotalBackendMetaMetrics(newBlockList, newCompactedBlockList)
-			metricBackendObjects.WithLabelValues(tenantID, blockStatusLiveLabel).Set(float64(backendMetaMetrics.blockMetaTotalObjects))
-			metricBackendObjects.WithLabelValues(tenantID, blockStatusCompactedLabel).Set(float64(backendMetaMetrics.compactedBlockMetaTotalObjects))
-			metricBackendBytes.WithLabelValues(tenantID, blockStatusLiveLabel).Set(float64(backendMetaMetrics.blockMetaTotalBytes))
-			metricBackendBytes.WithLabelValues(tenantID, blockStatusCompactedLabel).Set(float64(backendMetaMetrics.compactedBlockMetaTotalBytes))
-			continue
+			for consecutiveErrorsRemaining >= 0 {
+				newBlockList, newCompactedBlockList, err = p.pollTenantAndCreateIndex(ctx, tenantID, previous)
+				if err == nil {
+					break
+				}
+
+				consecutiveErrorsRemaining--
+			}
+
+			mtx.Lock()
+			defer mtx.Unlock()
+
+			if err != nil {
+				level.Error(p.logger).Log("msg", "failed to poll or create index for tenant", "tenant", tenantID, "err", err)
+				blocklist[tenantID] = previous.Metas(tenantID)
+				compactedBlocklist[tenantID] = previous.CompactedMetas(tenantID)
+
+				tenantFailures.Inc()
+				tenantFailuresRemaining--
+				finalErr.Store(err)
```

Comment:
instead of storing and returning only 1 error, let's just return a generic error on line 239.
```
	if tenantFailures.Load() > int32(p.cfg.TolerateTenantFailures) {
		return errors.New("too many tenant failures. abandoning polling cycle")
	}
```
just returning a single error could be misleading. if we return this one error we can review the logs and see all the errors that caused this method to fail.

Reply:
I pushed a commit to clean this up a bit and return a generic error.